### ISSUE TITLE
Data Imbalance  in Finetune Systemic Diseases Prediction

### ISSUE THREAD

**Issue by Linlinpei** [rmaphoh/RETFound_MAE#36]
Thank you very much for your meaningful work.
I am encountering a significant data imbalance issue while fine-tuning the Systemic Diseases Prediction downstream task. This imbalance is resulting in low performance metrics, specifically AUC, F1-score, and MCC. I would like to know if the authors have experienced similar challenges when fine-tuning the Systemic Diseases Prediction task and how they addressed the data imbalance problem.

**Comment by rmaphoh** [comment]
Hi, can you provide more details, e.g., how many cases are there for the minor categories?

**Comment by Salmanshams67** [comment]
ValueError: Only one class present in y_true. ROC AUC score is not defined in that case.

For OCTID.

**Comment by Linlinpei** [comment]
The dataset I’m using has a 1:4 ratio of positive to negative cases, with each patient containing 128 OCT images, so the positive class consists of approximately 220 * 128 images. Also, I’d like to ask, in your study, when dealing with an OCT sequence from a participant, do you choose a single OCT image or use all the images?

**Comment by Salmanshams67** [comment]
All the images.

**Comment by MDSALMANSHAMS** [comment]
### Issue:
**When running the command:** 
`python main_finetune.py --batch_size 1 --epochs 10 --model RETFound_mae --data_path OCTID`

**Following error was raised during evaluation:** 
`ValueError: Only one class present in y_true. ROC AUC score is not defined in that case.`

**Cause:** The error occurs in scikit-learn’s roc_auc_score function when it detects that one or more classes have no positive instances in the ground truth. In our case, the script was using the **default value --nb_classes 8**. However, the **OCTID** dataset is organized into **5 classes** (e.g., ANormal, ARMD, CSR, Diabetic_retinopathy, Macular_Hole).
Since ImageFolder assigns class indices alphabetically, the validation set contained labels for classes 0 through 4 only. The evaluation function then attempted to compute ROC AUC across 8 classes, resulting in an error because classes 5, 6, and 7 had no samples.

**Resolution:** The issue was resolved by specifying the correct number of classes for the OCTID dataset. Running the command with the flag **--nb_classes 5** ensures that the evaluation routine only expects 5 classes, thus avoiding the error: 
`python main_finetune.py --batch_size 1 --epochs 10 --model RETFound_mae --data_path OCTID --finetune checkpoint-best.pth --nb_classes 5`

**Notes:** Always verify that the --nb_classes parameter matches the actual class structure of your dataset.
This error is generated by scikit-learn when computing ROC AUC in a one-vs-rest (OVR) setting and is a good indicator that the dataset's label distribution does not match the expected number of classes.


@rmaphoh , please verify and close this issue.

Thanks,
Salman

**Comment by rmaphoh** [comment]
> ### Issue:
> **When running the command:** `python main_finetune.py --batch_size 1 --epochs 10 --model RETFound_mae --data_path OCTID`
> 
> **Following error was raised during evaluation:** `ValueError: Only one class present in y_true. ROC AUC score is not defined in that case.`
> 
> **Cause:** The error occurs in scikit-learn’s roc_auc_score function when it detects that one or more classes have no positive instances in the ground truth. In our case, the script was using the **default value --nb_classes 8**. However, the **OCTID** dataset is organized into **5 classes** (e.g., ANormal, ARMD, CSR, Diabetic_retinopathy, Macular_Hole). Since ImageFolder assigns class indices alphabetically, the validation set contained labels for classes 0 through 4 only. The evaluation function then attempted to compute ROC AUC across 8 classes, resulting in an error because classes 5, 6, and 7 had no samples.
> 
> **Resolution:** The issue was resolved by specifying the correct number of classes for the OCTID dataset. Running the command with the flag **--nb_classes 5** ensures that the evaluation routine only expects 5 classes, thus avoiding the error: `python main_finetune.py --batch_size 1 --epochs 10 --model RETFound_mae --data_path OCTID --finetune checkpoint-best.pth --nb_classes 5`
> 
> **Notes:** Always verify that the --nb_classes parameter matches the actual class structure of your dataset. This error is generated by scikit-learn when computing ROC AUC in a one-vs-rest (OVR) setting and is a good indicator that the dataset's label distribution does not match the expected number of classes.
> 
> [@rmaphoh](https://github.com/rmaphoh) , please verify and close this issue.
> 
> Thanks, Salman

Yes, this is correct. Aligning nb_classes to the category number of the application data is essential.

**Comment by laleoarrow** [comment]
Hi, so is there any suggestion for training a imbalanced data yet?
